recipes/CMIP6_static_grids/recipe.py

```diff
@@ -1,35 +1,48 @@
 from pangeo_forge_recipes.patterns import ConcatDim, FilePattern
 from pangeo_forge_recipes.recipes.xarray_zarr import XarrayZarrRecipe
 
+url_dict = {
+    "GFDL-ESM4":
+    "{mock_concat}ftp://ftp.gfdl.noaa.gov/perm/Alistair.Adcroft/MOM6-testing/OM4_05/ocean_hgrid.nc",
+    # "another_grid_key":
+    # "{mock_concat}ftp://another_grid_source_file_url",
+}
 
-# This seems cumbersome and not too elegant.
-# Is there a way to wrap the values of `url_dict` into functions programmatically?
-def gfdl_esm4_url():
-    return "ftp://ftp.gfdl.noaa.gov/perm/Alistair.Adcroft/MOM6-testing/OM4_05/ocean_hgrid.nc"
+time_concat_dim = ConcatDim("mock_concat", [""], nitems_per_file=1)
 
 
-url_dict = {"GFDL-ESM4": gfdl_esm4_url}
+def make_full_path(mock_concat, grid_key="default"):
+    return url_dict[grid_key].format(mock_concat=mock_concat)
 
 
 # Preprocessors
 
-# At the moment only a dummy func
-def gfdl_pp_func(ds):
-    ds.attrs["just_a_dummy_entru"] = "again just a dummy"
-    return ds
 
+def gfdl_pp_func(ds, fname):
+    return ds.drop("tile")  # dtype="|S255" is invalid for xarray
 
-preprocess_dict = {"GFDL-ESM4": gfdl_pp_func}
 
+preprocess_dict = {
+    "GFDL-ESM4": gfdl_pp_func,
+    # "another_grid_key": "another_grid_pp_func"
+}
 
-time_concat_dim = ConcatDim("dummy", list(range(0)))
+xr_open_kwargs_dict = {
+    "GFDL-ESM4": {"engine": "scipy"},
```

Comment:
Why would we use the scipy engine here?

Reply:
Because for some reason the default engine of `"h5netcf"` can't open this file.

Reply:
[This](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/e1ef5756ced85e5f9b029e1c40291e133a73329b/pangeo_forge_recipes/recipes/xarray_zarr.py#L204-L205) is where `"h5netcdf"` is set as default, but ...

```python
import xarray as xr
# below, 'ocean_hgrid.nc' is just the grid file, downloaded to my local drive
ds = xr.open_dataset('ocean_hgrid.nc', engine="h5netcdf")
print(ds)
```
returns ...

<details>
    <summary> Traceback </summary>

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/file_manager.py in _acquire_with_cache_info(self, needs_lock)
    198             try:
--> 199                 file = self._cache[self._key]
    200             except KeyError:

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/lru_cache.py in __getitem__(self, key)
     52         with self._lock:
---> 53             value = self._cache[key]
     54             self._cache.move_to_end(key)

KeyError: [<class 'h5netcdf.core.File'>, ('/home/jovyan/CMIP6_grids/ocean_hgrid.nc',), 'r', (('invalid_netcdf', None),)]

During handling of the above exception, another exception occurred:

OSError                                   Traceback (most recent call last)
<ipython-input-4-24c326657250> in <module>
      1 import xarray as xr
----> 2 ds = xr.open_dataset('ocean_hgrid.nc', engine="h5netcdf")
      3 print(ds)

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/api.py in open_dataset(filename_or_obj, group, decode_cf, mask_and_scale, decode_times, autoclose, concat_characters, decode_coords, engine, chunks, lock, cache, drop_variables, backend_kwargs, use_cftime, decode_timedelta)
    570 
    571         opener = _get_backend_cls(engine)
--> 572         store = opener(filename_or_obj, **extra_kwargs, **backend_kwargs)
    573 
    574     with close_on_error(store):

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/h5netcdf_.py in open(cls, filename, mode, format, group, lock, autoclose, invalid_netcdf, phony_dims)
    163 
    164         manager = CachingFileManager(h5netcdf.File, filename, mode=mode, kwargs=kwargs)
--> 165         return cls(manager, group=group, mode=mode, lock=lock, autoclose=autoclose)
    166 
    167     def _acquire(self, needs_lock=True):

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/h5netcdf_.py in __init__(self, manager, group, mode, lock, autoclose)
    105         # todo: utilizing find_root_and_group seems a bit clunky
    106         #  making filename available on h5netcdf.Group seems better
--> 107         self._filename = find_root_and_group(self.ds)[0].filename
    108         self.is_remote = is_remote_uri(self._filename)
    109         self.lock = ensure_lock(lock)

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/h5netcdf_.py in ds(self)
    174     @property
    175     def ds(self):
--> 176         return self._acquire()
    177 
    178     def open_store_variable(self, name, var):

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/h5netcdf_.py in _acquire(self, needs_lock)
    166 
    167     def _acquire(self, needs_lock=True):
--> 168         with self._manager.acquire_context(needs_lock) as root:
    169             ds = _nc4_require_group(
    170                 root, self._group, self._mode, create_group=_h5netcdf_create_group

/srv/conda/envs/notebook/lib/python3.8/contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/file_manager.py in acquire_context(self, needs_lock)
    185     def acquire_context(self, needs_lock=True):
    186         """Context manager for acquiring a file."""
--> 187         file, cached = self._acquire_with_cache_info(needs_lock)
    188         try:
    189             yield file

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/file_manager.py in _acquire_with_cache_info(self, needs_lock)
    203                     kwargs = kwargs.copy()
    204                     kwargs["mode"] = self._mode
--> 205                 file = self._opener(*self._args, **kwargs)
    206                 if self._mode == "w":
    207                     # ensure file doesn't get overriden when opened again

/srv/conda/envs/notebook/lib/python3.8/site-packages/h5netcdf/core.py in __init__(self, path, mode, invalid_netcdf, phony_dims, **kwargs)
    670                 else:
    671                     self._preexisting_file = os.path.exists(path)
--> 672                     self._h5file = h5py.File(path, mode, **kwargs)
    673             else:  # file-like object
    674                 if h5py.__version__ < LooseVersion('2.9.0'):

/srv/conda/envs/notebook/lib/python3.8/site-packages/h5py/_hl/files.py in __init__(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, **kwds)
    422             with phil:
    423                 fapl = make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0, **kwds)
--> 424                 fid = make_fid(name, mode, userblock_size,
    425                                fapl, fcpl=make_fcpl(track_order=track_order, fs_strategy=fs_strategy,
    426                                fs_persist=fs_persist, fs_threshold=fs_threshold),

/srv/conda/envs/notebook/lib/python3.8/site-packages/h5py/_hl/files.py in make_fid(name, mode, userblock_size, fapl, fcpl, swmr)
    188         if swmr and swmr_support:
    189             flags |= h5f.ACC_SWMR_READ
--> 190         fid = h5f.open(name, flags, fapl=fapl)
    191     elif mode == 'r+':
    192         fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5f.pyx in h5py.h5f.open()

OSError: Unable to open file (file signature not found)
```

</details>

Reply:
does `netcdf4` work?

Reply:
> does `netcdf4` work?

Reading from a local filesystem, yes it does. But `pangeo-forge-recipes` typically caches inputs to cloud storage, and this engine doesn't work for reading from Google Cloud Storage, for some reason:

```python
import xarray as xr

cached_file = (
    'pangeo-forge-us-central1/pangeo-forge-cache/cmip6_grids/GFDL-ESM4/'
    '0fbcf743691eeb06bf2bbef087a0d6d5-'
    'ftp_ftp.gfdl.noaa.gov_perm_alistair.adcroft_mom6-testing_om4_05_ocean_hgrid.nc'
)
f = fs_gcs.open(cached_file, mode="rb")
ds = xr.open_dataset(f, engine="netcdf4")
ds
```

<details>
     <summary> Traceback </summary>

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-19-dff2e73b0564> in <module>
      8 )
      9 f = fs_gcs.open(cached_file, mode="rb")
---> 10 ds = xr.open_dataset(f, engine="netcdf4")
     11 ds

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/api.py in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, backend_kwargs, *args, **kwargs)
    494 
    495     overwrite_encoded_chunks = kwargs.pop("overwrite_encoded_chunks", None)
--> 496     backend_ds = backend.open_dataset(
    497         filename_or_obj,
    498         drop_variables=drop_variables,

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/netCDF4_.py in open_dataset(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta, group, mode, format, clobber, diskless, persist, lock, autoclose)
    547 
    548         filename_or_obj = _normalize_path(filename_or_obj)
--> 549         store = NetCDF4DataStore.open(
    550             filename_or_obj,
    551             mode=mode,

/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/netCDF4_.py in open(cls, filename, mode, format, group, clobber, diskless, persist, lock, lock_maker, autoclose)
    351 
    352         if not isinstance(filename, str):
--> 353             raise ValueError(
    354                 "can only read bytes or file-like objects "
    355                 "with engine='scipy' or 'h5netcdf'"

ValueError: can only read bytes or file-like objects with engine='scipy' or 'h5netcdf'
```

</details>